### DISCUSSION TITLE
Spaces around = in PSR12 rule set (and other binary operators)

### DISCUSSION THREAD

**Discussion by duboism** [PHP-CS-Fixer/PHP-CS-Fixer#6849] [Q&A] [answered]
Hello,

I'm new to PHP-CS-Fixer so I'm sorry if the question is trivial (we searched but could not find the answer and a previous question on Gitter was left unanswered).

So I want to use PHP-CS-Fixer to convert an old code base to PSR-12. `php-cs-fixer fix --rules=@PSR12` worked like a charm but I have noticed that this doesn't insert spaces around `=` (other operators seems OK) while it's required in PSR-12. I know I can add the `binary_operator_spaces` rule as in #3959 but I was wondering why it's not part of the `@PSR12` rule set.

EDIT: I realized later that all binary operators are concerned, not just `=`.

**Comment by kubawerlos** [comment]
PSR12 requires:

> All binary [arithmetic](http://php.net/manual/en/language.operators.arithmetic.php), [comparison](http://php.net/manual/en/language.operators.comparison.php), [assignment](http://php.net/manual/en/language.operators.assignment.php), [bitwise](http://php.net/manual/en/language.operators.bitwise.php), [logical](http://php.net/manual/en/language.operators.logical.php), [string](http://php.net/manual/en/language.operators.string.php), and [type](http://php.net/manual/en/language.operators.type.php) operators MUST be preceded and followed by at least one space

and there is no config to achieve that in `binary_operator_spaces`.

**Comment by duboism** [comment]
Sorry, I don't understand your answer. I know that PSR12 requires spaces around `=` (and others). The problem is that AFAIU the `@PSR12` rule set doesn't enforce that by default:

```bash
$ cat test.php 
<?php

$a=42;  // No space
$ php-cs-fixer fix --dry-run --diff --rules=@PSR12 test.php 
Loaded config default.
Using cache file ".php-cs-fixer.cache".

Found 0 of 1 files that can be fixed in 0.002 seconds, 12.000 MB memory used
$ php-cs-fixer fix --dry-run --diff --rules='{"@PSR12": true ,"binary_operator_spaces": {"default": "single_space"}}' test.php 
Loaded config default.
Using cache file ".php-cs-fixer.cache".
   1) test.php
      ---------- begin diff ----------
--- /tmp/test.php
+++ /tmp/test.php
@@ -1,3 +1,3 @@
 <?php
 
-$a=42;  // No space
+$a = 42;  // No space

      ----------- end diff -----------


Found 1 of 1 files that can be fixed in 0.034 seconds, 12.000 MB memory used
```

Am I doing something wrong ? We use php-cs-fixer 3.14.4 on PHP 7.4.33.

**Comment by kubawerlos** [comment]
Consider a code:
```php
<?php

$foo = 1;
$b   = 2;
```
Is this correct PSR-12 code or not?

**Reply by duboism** [reply]
AFAIU, it is.

**Comment by kubawerlos** [comment, marked as answer]
Exactly, and your proposal:
`php-cs-fixer fix --dry-run --diff --rules='{"@PSR12": true ,"binary_operator_spaces": {"default": "single_space"}}'`
would change it, while current `@PSR12` ruleset does not, which is correct.

**Reply by duboism** [reply]
OK. But in the example I gave, there is no space at all around `=` so the default could be to include `binary_operator_spaces` (for instance with `single_space`).

**Reply by kubawerlos** [reply]
No, it cannot, because it will also break correctly PSR-12 code. The would need to be a new option, something like `at_least_single_space`.

**Reply by duboism** [reply]
I agree on that (I don't have enough experience with `php-cs-fixer` to realize that beforehand). Do you think it's worth implementing that (and maybe add it to `@PSR12`) ?

**Reply by kubawerlos** [reply]
Definitely worth it, if you brave enough, the current list of options (https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.15.1/doc/rules/operator/binary_operator_spaces.rst#default) scares me.

**Reply by duboism** [reply]
If you're scared, imagine how I feel ! More seriously, I took a look at the code and it would take me a lot of time to understand and modify it. Maybe I can open an issue and someone more competent could implement that ?

In any case, I understand the motivation behind the rules and I can work with that for now. So thanks for your time and explanations.

**Reply by Wirone** [reply]
Now, after discussion, when it's clear what has to be done, issue can be created 🙂. Discussions are better when action to take is not clear, that's why I often move issues to discussions to find the actual outcome.

**Reply by duboism** [reply]
OK, I will open the issue then. Note that, contrary to what I said, all binary operators are concerned , not just `=` (I think my coworkers and I got confused at some point). I will clarify that in the first message in case some people read this.

**Reply by duboism** [reply]
I opened #6861 (mentioning this discussion). I think we can close the discussion now. Thanks everyone for the discussion.
